common/lib/xmodule/setup.py

```diff
@@ -34,6 +34,7 @@
     "course_info = xmodule.html_module:CourseInfoBlock",
     "html = xmodule.html_module:HtmlBlock",
     "library = xmodule.library_root_xblock:LibraryRoot",
+    "library_sourced = xmodule.library_sourced_block:LibrarySourcedBlock",
```

Comment:
Apologies if I'm missing history here; why are we including this XBlock directly in the platform, instead of as its own package?

I noticed a note on the WIP PR:

> Note that once we get courses implemented in Blockstore as well I expect we won't use this block at all, so I haven't designed it to be forwards compatible in that sense.

which strikes me as all the more reason to keep as much of it as we can out of the core :)

And do we expect the broader community to start using this XBlock?

Should we at least flag the "to be unused" sections as such?
And/or create a relevant DEPR/follow-up ticket?

I'm generally uneasy about merging "temporary" work into the core, as "temporary" has a habit of hanging around ;)

I'm sure we have the best of intentions here, so as they say:

> The road to hell is paved with good intentions.

:)

Anyway, I'm curious.

Reply:
Good questions :)

There is a slightly related discussion at https://discuss.openedx.org/t/running-blockstore-in-juniper/2625/9 which may provide helpful context.

> why are we including this XBlock directly in the platform, instead of as its own package?

I consider "being able to use library content in a course" to be core functionality. Due to how the split modulestore and XBlock co-evolved over time, the implementation of that for v2 libraries is currently split between this XBlock and the "Library Tools Service" (which reaches into modulestore to do the dirty work). But, related to the [ADR for the role of XBlock](https://github.com/edx/edx-platform/pull/24538), "being able to use library content in a course" is not really something that should be built at the XBlock level, so once we implement course outlines and course content in Blockstore, this XBlock won't be relevant because you'll import content from a library by editing the course outline, which won't involve XBlocks at all.

So: if we want to allow users to use Libraries v2 in their [modulestore] courses during the transition period when we have blockstore libraries but not blockstore courses, then we need this XBlock (and a similar version that does randomization of the content).

A lot of the integration here happens in the "library service", so it's true that the XBlock itself could be moved into its own package and would have a relatively clean API boundary; essentially the separate package would just be this file and its tests: https://github.com/edx/edx-platform/blob/47f87b19ad2b57f0e81918664de5c53488068853/common/lib/xmodule/xmodule/library_sourced_block.py

However, the XBlock code is highly coupled to the "library service" that's in the core, it provides functionality that I consider core, and there is no package/software other than edxapp/edx-platform that would conceivably need to install it nor could even run it. We could move it to a separate repo and have it installed+tested by default, but that just adds a bunch of overhead and requires working across two repos in exchange for (imho) no benefit at all.

> And do we expect the broader community to start using this XBlock?

Yes. Personally I'd like to get this out there ASAP (along with a version that does randomization) and get as much content as possible moved into Blockstore libraries so we can deprecate v1/modulestore libraries sooner rather than later.

> And/or create a relevant DEPR/follow-up ticket?

This can't be deprecated before blockstore-based courses are implemented (a long time from now) and will probably need to stay for as long as modulestore-based courses do. It can be deprecated at the same time as modulestore.

> I'm generally uneasy about merging "temporary" work into the core, as "temporary" has a habit of hanging around ;)

I consider it more of an adapter (make the new libraries work with the old courses) that we'll need as long as we have old courses around, more so than a temporary thing.

Reply:
Thanks @bradenmacdonald !
I think I was mostly set off by the "once we get courses implemented in Blockstore as well I expect we won't use this block at all" comment.

Thanks for the clarity!

Reply:
Thanks for prompting me to think about this more :)